File: codechef/contests/SEPT16/CHFNFRN.cpp
/*input
1
6 6
2 6
1 5
5 6
4 5
1 3
2 4
*/

#include <bits/stdc++.h>

using namespace std;

#define inf INT_MAX
#define pii pair<int, int>
#define vi vector<int>
#define vvi vector<vi>
#define si set<int>
#define mii map<int, int>
#define mci map<char, int>
#define mic map<int, char>
#define vit vi::iterator
#define sit si::iterator
#define psb push_back
#define ppb pop_back
#define mp make_pair
#define init(a, n) a.find(n) != a.end()
#define F(i,n) for (int i = 0; i < n; i += 1)
#define inp(a,n) F(i,n) cin>>a[i]
#define print(a,n) F(i,n) {cout << a[i] << " ";} cout << endl
// graphs all will be 1 based indexed
#define uduwginp(g, m) F(i,m){cin>>x>>y;g[x].psb(y);g[y].psb(x);}
#define duwginp(g,m) F(i,m){cin>>x>>y; g[x].psb(y);}
#define printgraph(g,n) for(int i = 1; i <= n; i += 1){cout<< endl<<i<<": "; for(int adj:g[i]) cout << adj <<" ";} cout << endl

int **ispair, n, m;
vector<pii> a;

bool contradicting(vi temp){
	for(int s: temp){
		for (int t: temp){
			if(s != t and !ispair[s][t]){
				return true;
			}
		}
	}
	return false;
}

bool solve(vi first, vi second, si vis){
	if(vis.size() == a.size()) return true;
	// cout << "Solving" << endl;
	first.psb(a[0].second);
	vis.insert(0);
	for (int i = 1; i < a.size(); i += 1){
		if (!ispair[a[0].second][a[i].second]){
			second.psb(a[i].second);
			vis.insert(i);
		}
	}
	// cout << "Initial first: "; print(first, first.size());
	// cout << "Initial second: "; print(second, second.size());
	if(contradicting(second)) return false;
	int prev_size = vis.size();
	while (vis.size() < a.size()){
		// cout << "vis size: " << vis.size() << " a size: " << a.size() << endl;
		for (int i = 0; i < a.size(); i += 1){
			bool f1 = true, f2 = true;
			if (init(vis, i) == false){
				for (int s: first){
					if (!ispair[s][a[i].second]){
						f1 = false;
						break;
					}
				}
				for (int s: second){
					if (!ispair[s][a[i].second]){
						f2 = false;
						break;
					}
				}
				if(!f1 and !f2){
					return false;
				}else if(!f1){
					second.psb(a[i].second);
					vis.insert(i);
				}else if(!f2){
					first.psb(a[i].second);
					vis.insert(i);
				}
			}
		}
		if (vis.size() == prev_size){
			for(int i = 0; i < a.size(); i += 1){
				if(init(vis, i) == false){
					second.psb(a[i].second);
					vis.insert(i);
					break;
				}
			}
		}
		prev_size = vis.size();
	}
	return true;
}

int main(void){
	ios::sync_with_stdio(false);
	int t;
	cin >> t;
	while(t--){
		int x, y;
		// Input goes here
		cin >> n >> m;
		ispair = new int*[n+1];
		int degree[n+1];
		fill_n(degree, n+1, 0);
		for (int i = 0; i <= n; i += 1){
			ispair[i] = new int[n+1];
			for (int j = 0; j <= n; j +=1){
				ispair[i][j] = 0;
			}
		}
		for (int i = 0; i < m; i += 1){
			cin >>  x >> y;
			ispair[x][y] = 1;
			ispair[y][x] = 1;
			degree[x] += 1;
			degree[y] += 1;
		}
		// Input ends here
		// Solution starts here
		a.clear();
		for (int i = 1; i < n+1; i += 1){
			if(degree[i] < n-1)
				a.psb({degree[i], i});
		}
		vi first, second;
		sort(a.begin(), a.end());
		bool ans;
		if (a.empty()){
			ans = true;
		}else{
			si vis;
			ans = solve(first, second, vis);
		}
		if (ans){
			cout << "YES" << endl;
		}else{
			cout << "NO" << endl;
		}
	}
	
}

File: codechef/contests/SEPT16/CHFNFRN.py
'''input
1
6 6
2 6
1 5
5 6
4 5
1 3
2 4
'''

import os, sys, collections, itertools, fractions, math, random

ai = lambda : map(int, raw_input().split())
ii = lambda : int(raw_input())

degree = None
ispair = None
n = None
m = None
a = None

def comp(x, y):
	return degree[x] <= degree[y]

def contra(temp):
	for s in temp:
		for t in temp:
			if s != t and not ispair[s][t]:
				return 1
	return 0

def solve(first, second, vis):
	if len(vis) == len(a):
		return 1
	if not first and not second:
		first.append(a[0])
		vis.add(0)
		for i in xrange(1,len(a)):
			if (not ispair[a[0]][a[i]]):
				second.append(a[i])
				vis.add(i)
		if contra(second):
			return 0
	psize = len(vis)
	while len(vis) < len(a):
		for i in xrange(len(a)):
			f1, f2 = 1, 1
			if i not in vis:
				for s in first:
					if not ispair[s][a[i]]:
						f1 = 0
						break
				for s in second:
					if not ispair[s][a[i]]:
						f2 = 0
						break
				if not f1 and not f2:
					return 0
				elif not f1:
					second.append(a[i])
					vis.add(i)
				elif not f2:
					first.append(a[i])
					vis.add(i)
		if psize == len(vis):
			for i in xrange(len(a)):
				if i not in vis:
					# print "Hell yeah! I took the recursive calls"
					first.append(a[i])
					vis.add(i)
					fs = solve(first, second, vis)
					first.pop()
					second.append(a[i])
					ss = solve(first, second, vis)
					vis.remove(i)
					return (fs or ss)
		psize = len(vis)
	return 1



inputs = sys.stdin.read().split("\n");
ind = 0
t = int(inputs[ind])
ind += 1

for _ in xrange(t):
	n, m = map(int, inputs[ind].split())
	ind += 1
	ispair = [[0 for ii in xrange(n+1)] for jj in xrange(n+1)]
	degree = [0 for ii in xrange(n+1)]
	for __ in xrange(m):
		x, y = map(int, inputs[ind].split())
		ind += 1
		ispair[x][y] = 1
		ispair[y][x] = 1
		degree[x] += 1
		degree[y] += 1
	a = []
	for i in xrange(1,n+1):
		if degree[i] < n-1:
			a.append(i)
	if not a:
		ans = 1
	else:
		vis = set()
		a.sort(comp)
		fst = []
		sec = []
		ans = solve(fst, sec, vis)

	if ans: print "YES"
	else: print "NO"